Shell/Builtin.cpp

```diff
@@ -0,0 +1,723 @@
+            builder.append(cwd);
+            builder.append('/');
+            builder.append(arg_path);
+            new_path = builder.to_string();
+        }
+    }
+
+    auto real_path = Core::File::real_path_for(new_path);
+    if (real_path.is_empty()) {
+        fprintf(stderr, "Invalid path '%s'\n", new_path.characters());
+        return 1;
+    }
+    const char* path = real_path.characters();
+
+    int rc = chdir(path);
+    if (rc < 0) {
+        if (errno == ENOTDIR) {
+            fprintf(stderr, "Not a directory: %s\n", path);
+        } else {
+            fprintf(stderr, "chdir(%s) failed: %s\n", path, strerror(errno));
+        }
+        return 1;
+    }
+    setenv("OLDPWD", cwd.characters(), 1);
+    cwd = real_path;
+    setenv("PWD", cwd.characters(), 1);
+    return 0;
+}
+
+int Shell::builtin_cdh(int argc, const char** argv)
+{
+    int index = -1;
+
+    Core::ArgsParser parser;
+    parser.add_positional_argument(index, "Index of the cd history entry (leave out for a list)", "index", Core::ArgsParser::Required::No);
+
+    if (!parser.parse(argc, const_cast<char**>(argv), false))
+        return 1;
+
+    if (index == -1) {
+        if (cd_history.is_empty()) {
+            fprintf(stderr, "cdh: no history available\n");
+            return 0;
+        }
+
+        for (ssize_t i = cd_history.size() - 1; i >= 0; --i)
+            printf("%lu: %s\n", cd_history.size() - i, cd_history.at(i).characters());
+        return 0;
+    }
+
+    if (index < 1 || (size_t)index > cd_history.size()) {
+        fprintf(stderr, "cdh: history index out of bounds: %d not in (0, %zu)\n", index, cd_history.size());
+        return 1;
+    }
+
+    const char* path = cd_history.at(cd_history.size() - index).characters();
+    const char* cd_args[] = { "cd", path, nullptr };
+    return Shell::builtin_cd(2, cd_args);
+}
+
+int Shell::builtin_dirs(int argc, const char** argv)
+{
+    // The first directory in the stack is ALWAYS the current directory
+    directory_stack.at(0) = cwd.characters();
+
+    bool clear = false;
+    bool print = false;
+    bool number_when_printing = false;
+
+    Vector<const char*> paths;
+
+    Core::ArgsParser parser;
+    parser.add_option(clear, "Clear the directory stack", "clear", 'c');
+    parser.add_option(print, "Print directory entries", "print", 'p');
+    parser.add_option(number_when_printing, "Number the directories in the stack when printing", "number", 'v');
+    parser.add_positional_argument(paths, "Extra paths to put on the stack", "path", Core::ArgsParser::Required::No);
+
+    if (!parser.parse(argc, const_cast<char**>(argv), false))
+        return 1;
+
+    // -v implies -p
+    print = print || number_when_printing;
+
+    if (clear) {
+        for (size_t i = 1; i < directory_stack.size(); i++)
+            directory_stack.remove(i);
+    }
+
+    for (auto& path : paths)
+        directory_stack.append(path);
+
+    if (print || (!clear && paths.is_empty())) {
```

Comment:
`-p` doesn't do what you think it does (not in GNU Bash or zsh, at least). It's not "print even if a path is specified" (in fact, specifying both `-p` and a path should be an error), it simply changes the directory separator (`\n` instead of the default ` `) if we're already printing.

Reply:
hmmm...alright.
I don't think I've written the logic for this thing, so blame Andreas.

Reply:
@awesomekling casts **Absorb Blame (Rank 9)**

Reply:
Quick, use [`git blame-someone-else`](https://github.com/jayphelps/git-blame-someone-else) 😆

Reply:
There we go.
Specifies "print one per page" and fails when any path is given with -p or -v.

Reply:
@awesomekling it was super effective, or something like that.

Reply:
> use git blame-someone-else

I love it lol.